lib/primitives/address.js

```diff
@@ -148,13 +149,30 @@ class Address {
       case Address.types.SCRIPTHASH:
         return prefixes.scripthash;
       case Address.types.WITNESS:
-        if (this.hash.length === 20)
-          return prefixes.witnesspubkeyhash;
-
-        if (this.hash.length === 32)
-          return prefixes.witnessscripthash;
-
-        break;
+        if (this.version === 0) {
+          if (this.hash.length === 20)
+            return prefixes.witnesspubkeyhash;
+
+          if (this.hash.length === 32)
+            return prefixes.witnessscripthash;
+
+          // BIP141 defines any version 0 data lengths besides 20 and 32
+          // as invalid. Note that this is not the case with version 1
+          // (and presumably, all other future versions) where specific
+          // lengths are defined as valid and trigger extra rule sets,
+          // but all other lengths remain unencumbered (ANYONECANSPEND).
+          return -1;
+        }
+
+        // Since all segwit address strings have the same prefix
+        // for each network, there really isn't any reason why version 0
+        // script/pubkey should have been explicitly identified (above).
+        // We keep them for backwards comptability but from witness version 1
+        // onward, we only need to retun the witness version number.
+        // Unlike legacy addresses, the prefix is not actually included in the
+        // address string. bcoin uses it internally to keep Address objects
+        // consistent (the addrIndexer expects it for each entry, for example).
+        return prefixes.witnessVersionMask + this.version;
```

Comment:
Since addr indexer users need to reindex *anyway*, we could actually fix the tech debt here:
Remove `prefixes.witnesspubkeyhash` and `prefixes.witnessscripthash` entirely from bcoin and just use the witness version number, which (as the comment explains) makes more sense overall.